### ISSUE TITLE
AI day3

### ISSUE THREAD

**Issue by Romangduck** [Romangduck/Day-1-Linux#12]
#Validation 개념 

 

validation 데이터 없이 train / test로만 분리하게 되면

train으로 만든 모델을 검증하기 위해 우리는 test data만 사용할 것입니다.

이때, 고정된 test data만 가지고 모델의 성능을 확인하고, 파라미터를 수정합니다.

이 과정을 반복하게 되면 결국 최종 모델은 test data에만 성능이 좋은 모델이 될 것입니다.

즉, test data에 과적합(Overfit)이 되어 다른 unseen data가 들어왔을 때는 예측력이 떨어질 수 있습니다.

 

이때 나오는 개념이 Validation입니다. 

ex)_

​

# 3. 컴파일 , 훈련

​

model.compile(loss='mse',optimizer='adam')

​

start_time = time.time()

hist = model.fit(x_train,y_train , epochs=500, batch_size=200,

          validation_split=0.2) # validation date -> 0.2 (train0.6/test0.2)

end_time = time.time() - start_time

# 4. 평가 , 예측

​

loss = model.evaluate(x_test, y_test)   # loss :  0.6231313347816467

print('loss : ', loss)

​

y_predict = model.predict(x_test) # x 전체값 넣어주는거보단 나음!

r2 = r2_score(y_test, y_predict)

print('r2 스코어 :', r2)    # r2 스코어 : 0.5297377856444552

​

print(end_time)

​

컴파일 단계에서 들어감.

# 시각화

​

#시각화

import matplotlib.pyplot as plt

plt.figure(figsize=(9,6))

plt.plot(hist.history['loss'], marker='.' , c='blue',label='loss')

plt.plot(hist.history['val_loss'],marker='.',c='orange',label='val_loss')

plt.title('Loss & Val_loss')

plt.ylabel('loss')

plt.xlabel('epochs')

plt.legend()

plt.show()

​

Train / Validation / Test data 3가지는 각각 모델을 학습하고 검증하고 평가하는데 목적이 다릅니다. 

​

* Train *

- 모델을 학습하기 위한 데이터셋으로 이때 학습은 최적의 파라미터를 찾는 것입니다.

- 즉, Train data는 오직 학습을 위한 데이터셋 

 

* Validation * 

- 학습이 이미 완료된 모델을 검증하기 위한 데이터셋입니다. 

- 학습이 된 여러 가지 모델 중 가장 좋은 하나의 모델을 고르기 위한 데이터셋입니다. 

- 학습 과정에 어느 정도 관여를 한다고 볼 수 있습니다. 

​- 하지만 Validation 데이터 자체가 학습에 직접적으로 관여하는 것은 아님 

 

* test * 

- 모델의 '최종 성능'을 평가하기 위한 데이터셋 

- 학습 과정에 관여를 하지 않음

 

" Train으로 학습하고 Validation으로 검증하고 Test로 최종 성능을 평가한다 " 

​

Early Stopping

▪ 과적합을 방지하기 위한 방법 중 하나 

▪ 학습 중 모델의 성능을 모니터링하다가 학습 데이터에 대한 성능은 계속해서 향상되지만 검증 데이터에 대한 성능은 향상되지 않거나 감소하는 지점에서 학습을 조기 종료하는 것 

• 일반적으로, 학습 데이터와 검증 데이터를 나누어 학습하면서 적용됨

epochs 많이 사용하는것은 과적합이 나는 원인중 하나이다.

이를 방지하기위해,  epoch를 돌리는와중 특정 시점에서 멈추게 하는것이다.

여기서 특정시점은 일반적으로 훈련 중 주기적으로 성능 검증 과정중 성능이 더 좋아지지 않을때를 말한다.

​

# 3. 컴파일 , 훈련

​

model.compile(loss='mse',optimizer='adam')

​

##earlyStopping

from keras.callbacks import EarlyStopping

earlyStopping = EarlyStopping(monitor='val_loss', patience=100, mode='min',

                              verbose=1, restore_best_weights=True) ## restore_best_weights의 default는 False이므로 True로 *무조건 변경해줘야함.

​

​

start_time = time.time()

hist = model.fit(x_train,y_train , epochs=5000, batch_size=100,

          validation_split=0.2,

          callbacks=[earlyStopping],

          verbose=1) # validation date -> 0.2 (train0.6/test0.2)

end_time = time.time() - start_time

​

​

-> earlyStopping은 컴파일 훈련단계 , 컴파일 이후에 적어주자.

​

*restore_best_weights는 ModelCheckpoint 콜백 함수의 하나의 옵션으로, 모델의 최상의 성능을 보인 가중치를 자동으로 복원하는 기능입니다. 예를 들어, 검증 손실이나 검증 정확도가 최적값을 갱신했을 때마다 모델의 가중치를 저장할 수 있습니다. 이렇게 저장된 가중치는 후에 모델이 다시 로드될 때 restore_best_weights=True 옵션을 설정하여 자동으로 로드될 수 있습니다. 이를 통해 모델의 성능을 유지하면서 이전 상태로 되돌리는 것이 가능합니다.

​

​

data와 target을 시각화, 한글로

​

## loss 와 val_loss 시각화

import matplotlib.pyplot as plt

import os

os.environ['KMP_DUPLICATE_LIB_OK'] = 'True'

from matplotlib import font_manager, rc

font_path = 'C:/Windows\\Fonts/malgun.ttf'      #malgun은 폰트 이름, ttf는 파일 호환.

font = font_manager.FontProperties(fname=font_path).get_name()

rc('font',family=font)

plt.figure(figsize=(9, 6))

plt.plot(hist.history['loss'],marker='.',c='red',label='loss')

plt.plot(hist.history['val_loss'],marker='.',c='blue',label='val_loss')

plt.title('로스와 발로스')

plt.ylabel('loss')   #y축

plt.xlabel('epochs')    #x축

plt.legend()

plt.show()

​

로드와 세이브, Model Check Point

​

모델을 세이브할때 

​

model.save('./ _save/tf15_california.h5') #h5 로 모델 저장 -> model. 만 저장이됨,

                                          # time,hist와 같이 정의한 변수는 로드가 안됨, **model.fit ~ ~ 은 저장됨.

                                          #단지 hist만 저장이 안됨.

​

특이점 ; 모델만 저장됨, 당연히 컴파일 훈련 하기전 모델구성을 완성한후에 저장.

​

모델을 로드할때

import numpy as np

from keras.models import Sequential , load_model    # load_model 을 가져와야 실행가능

from keras.layers import Dense

from sklearn.model_selection import train_test_split

from sklearn.metrics import r2_score

# from sklearn.datasets import load_boston // 윤리적 문제로 제공 안됨

from sklearn.datasets import fetch_california_housing

datasets = fetch_california_housing()

import time

​

모델구성을 전부 지워버려도 모델구성 파일을 저장한것을 통으로 가져오기때문에 문제없이 진행가능.

# 2. 모델구성

​

# model = Sequential()

# model.add(Dense(100,input_dim=8))

# model.add(Dense(200))

# model.add(Dense(100))

# model.add(Dense(200))

# model.add(Dense(100))

# model.add(Dense(1))

​

#model.save('./ _save/tf15_california.h5') #h5 로 모델 저장

model = load_model('./ _save/tf15_california.h5')  #저장한 모델구성을 불러오는것 // 모델구성이 입력되지않아있음에도 돌아가는것을 확인할 수 있음

model = load_model('./ 파일 루트 ' ) 를 하면 불러옴

​

​

​

## Model Check Point

mcp = ModelCheckpoint(monitor='val_loss',

                      mode='auto',

                      verbose=1,

                      save_best_only=True,  

                      filepath='./_mcp/tf18_california.hdf5')

​

​

컴파일까지 모델 , 저장 가능 , 데이터입력란빼고 거의다 가능함 !
